breacharbiter_test.go

```diff
@@ -0,0 +1,398 @@
+		{
+			commitHash: [chainhash.HashSize]byte{
+				0xb7, 0x94, 0x38, 0x5f, 0x2d, 0x1e, 0xf7, 0xab,
+				0x4d, 0x92, 0x73, 0xd1, 0x90, 0x63, 0x81, 0xb4,
+				0x4f, 0x2f, 0x6f, 0x25, 0x88, 0xa3, 0xef, 0xb9,
+				0x6a, 0x49, 0x18, 0x83, 0x31, 0x98, 0x47, 0x53,
+			},
+			chanPoint:     breachOutPoints[0],
+			selfOutput:    &breachedOutputs[0],
+			revokedOutput: &breachedOutputs[1],
+			htlcOutputs:   []*breachedOutput{},
+		},
+		{
+			commitHash: [chainhash.HashSize]byte{
+				0x51, 0xb6, 0x37, 0xd8, 0xfc, 0xd2, 0xc6, 0xda,
+				0x48, 0x59, 0xe6, 0x96, 0x31, 0x13, 0xa1, 0x17,
+				0x2d, 0xe7, 0x93, 0xe4, 0xb7, 0x25, 0xb8, 0x4d,
+				0x1f, 0xb, 0x4c, 0xf9, 0x9e, 0xc5, 0x8c, 0xe9,
+			},
+			chanPoint:     breachOutPoints[1],
+			selfOutput:    &breachedOutputs[0],
+			revokedOutput: &breachedOutputs[1],
+			htlcOutputs: []*breachedOutput{
+				&breachedOutputs[1],
+				&breachedOutputs[2],
+			},
+		},
+	}
+)
+
+// Parse the pubkeys in the breached outputs.
+func initBreachedOutputs() error {
+	for i := 0; i < len(breachedOutputs); i++ {
+		bo := &breachedOutputs[i]
+
+		// Parse the sign descriptor's pubkey.
+		sd := &breachSignDescs[i]
+		pubkey, err := btcec.ParsePubKey(breachKeys[i], btcec.S256())
+		if err != nil {
+			return fmt.Errorf("unable to parse pubkey: %v", breachKeys[i])
+		}
+		sd.PubKey = pubkey
+		bo.signDescriptor = sd
+	}
+
+	return nil
+}
+
+// Test that breachedOutput Encode/Decode works.
+func TestBreachedOutputSerialization(t *testing.T) {
+	if err := initBreachedOutputs(); err != nil {
+		t.Fatalf("unable to init breached outputs: %v", err)
+	}
+
+	for i := 0; i < len(breachedOutputs); i++ {
+		bo := &breachedOutputs[i]
+
+		var buf bytes.Buffer
+
+		if err := bo.Encode(&buf); err != nil {
+			t.Fatalf("unable to serialize breached output [%v]: %v", i, err)
+		}
+
+		desBo := &breachedOutput{}
+		if err := desBo.Decode(&buf); err != nil {
+			t.Fatalf("unable to deserialize breached output [%v]: %v", i, err)
+		}
+
+		if !reflect.DeepEqual(bo, desBo) {
+			t.Fatalf("original and deserialized breached outputs not equal:\n"+
+				"original     : %+v\n"+
+				"deserialized : %+v\n",
+				bo, desBo)
+		}
+	}
+}
+
+// Test that retribution Encode/Decode works.
+func TestRetributionSerialization(t *testing.T) {
+	if err := initBreachedOutputs(); err != nil {
+		t.Fatalf("unable to init breached outputs: %v", err)
+	}
+
+	for i := 0; i < len(retributions); i++ {
+		ret := &retributions[i]
+
+		var buf bytes.Buffer
+
+		if err := ret.Encode(&buf); err != nil {
+			t.Fatalf("unable to serialize retribution [%v]: %v", i, err)
+		}
+
+		desRet := &retribution{}
+		if err := desRet.Decode(&buf); err != nil {
+			t.Fatalf("unable to deserialize retribution [%v]: %v", i, err)
+		}
+
+		if !reflect.DeepEqual(ret, desRet) {
+			t.Fatalf("original and deserialized retribution infos not equal:\n"+
+				"original     : %+v\n"+
+				"deserialized : %+v\n",
+				ret, desRet)
+		}
+	}
+}
+
+// TODO(phlip9): reuse existing function?
```

Comment:
If an interface is erected in front of the abstraction which persists the retribution info, then we won't need to also create a test database in order to write unittests and we gain the nice trait of not having the `breachArbiter` be closely coupled to `channeldb`.